hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-router/src/main/java/org/apache/hadoop/yarn/server/router/rmadmin/RouterRMAdminService.java

```diff
@@ -215,43 +190,24 @@ protected Map<String, RequestInterceptorChainWrapper> getPipelines() {
   @VisibleForTesting
   protected RMAdminRequestInterceptor createRequestInterceptorChain() {
     Configuration conf = getConfig();
+    RMAdminRequestInterceptor pipeline = null;
+    ClientMethod remoteMethod = null;
+    try {
+      remoteMethod = new ClientMethod("setNextInterceptor",
+              new Class[]{RMAdminRequestInterceptor.class}, new Object[]{null});
 
-    List<String> interceptorClassNames = getInterceptorClassNames(conf);
+      pipeline = RouterServerUtil.createRequestInterceptorChain(conf,
+          YarnConfiguration.ROUTER_RMADMIN_INTERCEPTOR_CLASS_PIPELINE,
+          YarnConfiguration.DEFAULT_ROUTER_RMADMIN_INTERCEPTOR_CLASS,
+          remoteMethod, RMAdminRequestInterceptor.class);
 
-    RMAdminRequestInterceptor pipeline = null;
-    RMAdminRequestInterceptor current = null;
-    for (String interceptorClassName : interceptorClassNames) {
-      try {
-        Class<?> interceptorClass = conf.getClassByName(interceptorClassName);
-        if (RMAdminRequestInterceptor.class
-            .isAssignableFrom(interceptorClass)) {
-          RMAdminRequestInterceptor interceptorInstance =
-              (RMAdminRequestInterceptor) ReflectionUtils
-                  .newInstance(interceptorClass, conf);
-          if (pipeline == null) {
-            pipeline = interceptorInstance;
-            current = interceptorInstance;
-            continue;
-          } else {
-            current.setNextInterceptor(interceptorInstance);
-            current = interceptorInstance;
-          }
-        } else {
-          throw new YarnRuntimeException(
-              "Class: " + interceptorClassName + " not instance of "
-                  + RMAdminRequestInterceptor.class.getCanonicalName());
-        }
-      } catch (ClassNotFoundException e) {
+      if (pipeline == null) {
         throw new YarnRuntimeException(
-            "Could not instantiate RMAdminRequestInterceptor: "
-                + interceptorClassName,
-            e);
+            "RequestInterceptor pipeline is not configured in the system.");
       }
-    }
-
-    if (pipeline == null) {
-      throw new YarnRuntimeException(
-          "RequestInterceptor pipeline is not configured in the system");
+    } catch (IOException | InvocationTargetException | NoSuchMethodException | RuntimeException
+             | IllegalAccessException ex) {
+      throw new YarnRuntimeException("Create RequestInterceptor Chain error.", ex);
     }
     return pipeline;
```

Comment:
Given that if error we always return exception, can we define the `pipeline` inside of the try and do the exception returns properly?

Reply:
Thank you for your suggestion. I handled different Exceptions in RouterServerUtil#createRequestInterceptorChain, and used YarnRuntimeException to make it easier for the caller to use it.